network/tx.go

```diff
@@ -0,0 +1,53 @@
+package network
+
+import (
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/crypto"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	"time"
+)
+
+// NewTransaction creates a new TransactionBuilder with the given required properties.
+func NewTransaction(payloadType string, payload []byte, key crypto.Key) TransactionBuilder {
```

Comment:
No need for this method, struct can be used directly.  This name is already in use by the DAG anyway.

Reply:
unless you want to enforce a minimum use. But then it would also require an interface and private struct

Reply:
It started with Codeclimate (rightly) complaining about the number of arguments of `CreateTransaction`. I considered the Options pattern, but that became ugly because it's not mapped to a single struct (but rather plucked apart and used as arguments for several underlying function calls). So I chose a sort of Builder pattern, but without the actual `Build()` step (and as you noticed, protected inners).

Reason is it can't build a `Transaction`, since that required multiple calls (new TX, sign it) and checks performed using dependencies (DAG to check the presence of all specified `prev`s, fetching DAG heads).  So I can't have a `TransactionBuilder` building a `Transaction` (or we could, but would require refactoring the complete transaction creation layer).

Most pure application of the pattern would be introducing a `TransactionSpec` (like a "blueprint") and having a `TransactionSpecBuilder`, which's `Build()` function would yield an immutable `TransactionSpec`. But that feels severly overengineered to me.

Would it be better if I unexported the fields of `TransactionBuilder` and renamed it to `TransactionTemplate`? Because it's not a true application of the Builder pattern, and more like a template with convenience methods.

Reply:
I was more triggered by the `New` in `NewTransaction` since that pattern usually returns an interface to an underlying (pointer to a) struct. You could just do: `TransactionTemplate{A, B, C}`? We have so many `New` and `Create` methods all over the place, that those start to lose meaning.

Reply:
I agree. Will remove the "New" method.